testinfra/utils/ansible_runner.py

```diff
@@ -18,150 +18,183 @@
+        if os.path.exists(data_dir):
+            shutil.rmtree(data_dir)
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir)
+
+        # runner must have an inventory file
+        inv_dir = os.path.join(data_dir, 'inventory')
+        if not os.path.exists(inv_dir):
+            os.makedirs(inv_dir)
+        shutil.copy(
+            self.host_list,
+            os.path.join(inv_dir, os.path.basename(self.host_list))
         )
 
-    def runner_on_skipped(self, host, item=None):
-        self.result = {
-            'failed': True,
-            'msg': 'Skipped. You might want to try check=False',
-            'item': item,
+        # molecule inventories use lookups
+        this_env = os.environ.copy()
+        this_env['ANSIBLE_NOCOLOR'] = "1"
+        env_keys = list(this_env.keys())
+        for ekey in env_keys:
+            if not ekey.startswith('MOLECULE'):
+                this_env.pop(ekey)
+        env_dir = os.path.join(data_dir, 'env')
+        if not os.path.exists(env_dir):
+            os.makedirs(env_dir)
+        with open(os.path.join(env_dir, 'envvars'), 'w') as f:
+            f.write('---\n')
+            f.write(yaml.dump(this_env))
+
+
+        # build the kwarg payload ansible-runner requires
+        runner_kwargs = {
+            'private_data_dir': data_dir,
+            'host_pattern': host,
+            'module': module_name,
+            'module_args': module_args,
+            'json_mode' : True,
         }
 
+        # ansible-runner does not have kwargs for these
+        for opt in ['become', 'check']:
+            if kwargs.get(opt, False):
+                if 'cmdline' not in runner_kwargs:
+                    runner_kwargs['cmdline'] = '--%s' % opt
+                else:
+                    runner_kwargs['cmdline'] += ' --%s' % opt
 
-class AnsibleRunner(AnsibleRunnerBase):
+        return AnsibleRunnerV2.call_runner(runner_kwargs, host)
 
-    def __init__(self, host_list=None):
-        super(AnsibleRunner, self).__init__(host_list)
-        self.cli = ansible.cli.playbook.PlaybookCLI(None)
-        self.cli.options = self.cli.base_parser(
-            connect_opts=True,
-            meta_opts=True,
-            runas_opts=True,
-            subset_opts=True,
-            check_opts=True,
-            inventory_opts=True,
-            runtask_opts=True,
-            vault_opts=True,
-            fork_opts=True,
-            module_opts=True,
-        ).parse_args([])[0]
-        self.cli.normalize_become_options()
-        self.cli.options.connection = "smart"
-        self.cli.options.inventory = host_list
-        # pylint: disable=protected-access
-        self.loader, self.inventory, self.variable_manager = (
-            self.cli._play_prereqs(self.cli.options))
+    @staticmethod
+    def call_runner(runner_kwargs, host):
+        # invoke ansible-runer -> ansible adhoc
+        r = ansible_runner.run(**runner_kwargs)
```

Comment:
``ansible_runner`` doesn't have a more performant mode for running commands with ansible ? Did you (or can you) compare test speeds before and after your PR ?

I wasn't aware of this license issue, I guess the path to have something performant is to include a kind of "ansible-server" in ansible and allow to run command through a pipe. This have been done in mercurial for similar license and performance reasons. Just wondering what is the state of this for ansible & ansible-runner ?

That being said, having a working maintainable runner and fix the license issue is more important than performances I think.

Reply:
Architecturally, runner has been designed for the needs of Ansible Tower/AWX where the primary payload is an inventory and a playbook and the session would never persist. I do note that this means pretty bad performance with many subsequent Ansible calls for scenarios like testinfra creates. If runner eventually gets a standard "server" implementation, I think it will still architecturally run with the same inventory+playbook idea in mind. However, that's not to say that testinfra / molecule's usecase couldn't push a new paradigm within runner and I personally have some hacky prototypes, but I think getting our head above water with what we've got now is the priority as you've already mentioned.